### PR TITLE
Serenity based framework

### PR DESCRIPTION
**Pull request by Invisib1eTouch**

### CODE REVIEW

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                .addPassword(password)
+                .build();
+
+        response = given()
+                .contentType(JSON)
+                .body(gson.toJson(loginModel))
+                .when()
+                .post(ENDPOINT);
+
+        token = generateToken(username, password);
```
Comment (JetFree) [inline src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: should be **this**.token = generateToken(username, password);

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        userID = response
```
Comment (JetFree) [inline src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: **this.**userID

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    public void successful_registration_verification() {
+        String givenUserID = given()
+                .header("Authorization", "Bearer " + token)
+                .contentType(JSON)
+                .when()
+                .get(ENDPOINT + "/" + userID)
+                .then()
+                .extract().jsonPath().get("userId");
+
+        assertThat(response.getStatusCode(), equalTo(HttpStatus.SC_CREATED));
```
Comment (JetFree) [inline src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: To be honest here, additionally, we need to verify that this account really exists and created

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        JsonPath resultJsonPath = response.then().extract().jsonPath();
+
+        assertThat(response.getStatusCode(), equalTo(HttpStatus.SC_BAD_REQUEST));
+        assertThat(resultJsonPath.get("code"), equalTo(errorCode));
+        assertThat(resultJsonPath.get("message"), equalTo(errorMessage));
+    }
+
+    @Then("User is not registered")
+    public void registration_with_incorrect_credentials_verification() {
+        assertThat(token, equalTo(null));
```
Comment (JetFree) [inline src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: It's not a correct check for this method. Token = null doesn't mean that user is not registered

File: src/main/java/apisteps/BookStoreApiSteps/StoreSteps.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                .when()
+                .get(BOOKS_ENDPOINT);
+    }
+
+    @When("Any book is added to user's collection")
+    public void add_book_to_user_collection() {
+        BooksModel booksInStore = gson.fromJson(response.then().extract().body().asString(), BooksModel.class);
+
+        for (int i = 0; i < AuthorizationSteps.userCredentialsModels.size(); i++) {
+            bookModels.add(booksInStore.getBooks()[DataGenerator.getRandomNumber(0, booksInStore.getBooks().length - 1)]);
```
Comment (JetFree) [inline src/main/java/apisteps/BookStoreApiSteps/StoreSteps.java]: just unreadable

File: src/main/java/models/BookModel.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    public String getWebsite() {
+        return website;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        BookModel bookModel = (BookModel) o;
+        return pages == bookModel.pages && Objects.equals(isbn, bookModel.isbn) && Objects.equals(title, bookModel.title) && Objects.equals(subTitle, bookModel.subTitle) && Objects.equals(author, bookModel.author) && Objects.equals(publishDate, bookModel.publishDate) && Objects.equals(publisher, bookModel.publisher) && Objects.equals(description, bookModel.description) && Objects.equals(website, bookModel.website);
```
Comment (JetFree) [inline src/main/java/models/BookModel.java]: Unreadable and awful, try to find a way how to compare all data that you need without such pain

File: src/main/java/models/BookModel.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(isbn, title, subTitle, author, publishDate, publisher, pages, description, website);
```
Comment (JetFree) [inline src/main/java/models/BookModel.java]: any reason why it should be overridden?

File: src/main/java/models/jwtToken/DecodedTokenHeader.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        public DecodedTokenHeader build() {
+            return decodedTokenHeader;
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (JetFree) [inline src/main/java/models/jwtToken/DecodedTokenHeader.java]: I need an explanation here. It is totally has no sense for me

**Review by JetFree — CHANGES_REQUESTED** [review]

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java (hunk 1 above)
Comment (Invisib1eTouch) [reply src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: Fixed.

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java (hunk 2 above)
Comment (Invisib1eTouch) [reply src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: Fixed.

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java (hunk 4 above)
Comment (Invisib1eTouch) [reply src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: There is no access to DB so I can’t verify user registration info. Current verification is based on the idea that user should be in DB to generate token. And when token is not generated it means that it doesn’t exist in DB.

File: src/main/java/apisteps/BookStoreApiSteps/StoreSteps.java (hunk 5 above)
Comment (Invisib1eTouch) [reply src/main/java/apisteps/BookStoreApiSteps/StoreSteps.java]: Fixed.

File: src/main/java/models/BookModel.java (hunk 6 above)
Comment (Invisib1eTouch) [reply src/main/java/models/BookModel.java]: Fixed.

File: src/main/java/models/BookModel.java (hunk 7 above)
Comment (Invisib1eTouch) [reply src/main/java/models/BookModel.java]: Removed.

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java (hunk 3 above)
Comment (Invisib1eTouch) [reply src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: There is no access to DB so I can’t verify that account really exists. That’s why I compare User IDs: User ID that received in response body after user creation request and User ID that received in get user details info request (means that request gave data from DB).

File: src/main/java/models/jwtToken/DecodedTokenHeader.java (hunk 8 above)
Comment (Invisib1eTouch) [reply src/main/java/models/jwtToken/DecodedTokenHeader.java]: Removed.

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java (hunk 3 above)
Comment (JetFree) [reply src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: ok, that's what I mean

File: src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java (hunk 4 above)
Comment (JetFree) [reply src/main/java/apisteps/BookStoreApiSteps/AccountSteps.java]: Here I assumed that you have possibility to get info about account through the API, there is no need to use DB

**Pull request closed without merging**